File: headers/carma/tests/src/test_type_caster.cpp
#include "test_type_caster.h"

namespace carma {
namespace tests {

double test_tc_in_mat(arma::Mat<double>& mat) {
    return arma::accu(mat);
}

double test_tc_in_row(arma::Row<double>& mat) {
    return arma::accu(mat);
}

double test_tc_in_col(arma::Col<double>& mat) {
    return arma::accu(mat);
}

double test_tc_in_cube(arma::Cube<double>& mat) {
    return arma::accu(mat);
}

arma::Mat<double> test_tc_out_mat(const py::array_t<double>& arr) {
    arma::Mat<double> ones = arma::ones(arr.shape(0), arr.shape(1));
    arma::Mat<double> mat = carma::arr_to_mat<double>(arr);
    arma::Mat<double> out = ones + mat;
    return out;
}

arma::Mat<double> test_tc_out_mat_rvalue(const py::array_t<double>& arr) {
    arma::Mat<double> ones = arma::ones(arr.shape(0), arr.shape(1));
    arma::Mat<double> mat = carma::arr_to_mat<double>(arr);
    return ones + mat;
}

arma::Row<double> test_tc_out_row(const py::array_t<double>& arr) {
    arma::Row<double> ones = arma::Row<double>(arr.size(), arma::fill::ones);
    arma::Row<double> mat = carma::arr_to_row<double>(arr);
    arma::Row<double> out = ones + mat;
    return out;
}

arma::Row<double> test_tc_out_row_rvalue(const py::array_t<double>& arr) {
    arma::Row<double> ones = arma::Row<double>(arr.size(), arma::fill::ones);
    arma::Row<double> mat = carma::arr_to_row<double>(arr);
    return ones + mat;
}

arma::Col<double> test_tc_out_col(const py::array_t<double>& arr) {
    arma::Col<double> ones = arma::Col<double>(arr.size(), arma::fill::ones);
    arma::Col<double> mat = carma::arr_to_col<double>(arr);
    arma::Col<double> out = ones + mat;
    return out;
}

arma::Col<double> test_tc_out_col_rvalue(const py::array_t<double>& arr) {
    arma::Col<double> ones = arma::Col<double>(arr.size(), arma::fill::ones);
    arma::Col<double> mat = carma::arr_to_col<double>(arr);
    return ones + mat;
}

arma::Cube<double> test_tc_out_cube(const py::array_t<double>& arr) {
    arma::Cube<double> ones = arma::Cube<double>(arr.shape(0), arr.shape(1), arr.shape(2), arma::fill::ones);
    arma::Cube<double> mat = carma::arr_to_cube<double>(arr);
    arma::Cube<double> out = ones + mat;
    return out;
}

arma::Cube<double> test_tc_out_cube_rvalue(const py::array_t<double>& arr) {
    arma::Cube<double> ones = arma::Cube<double>(arr.shape(0), arr.shape(1), arr.shape(2), arma::fill::ones);
    arma::Cube<double> mat = carma::arr_to_cube<double>(arr);
    return ones + mat;
}

}  // namespace tests
}  // namespace carma

void bind_test_tc_in_mat(py::module& m) {
    m.def("tc_in_mat", &carma::tests::test_tc_in_mat, "Test type caster");
}

void bind_test_tc_in_row(py::module& m) {
    m.def("tc_in_row", &carma::tests::test_tc_in_row, "Test type caster");
}

void bind_test_tc_in_col(py::module& m) {
    m.def("tc_in_col", &carma::tests::test_tc_in_col, "Test type caster");
}

void bind_test_tc_in_cube(py::module& m) {
    m.def("tc_in_cube", &carma::tests::test_tc_in_cube, "Test type caster");
}

void bind_test_tc_out_mat(py::module& m) {
    m.def("tc_out_mat", &carma::tests::test_tc_out_mat, "Test type caster");
}

void bind_test_tc_out_mat_rvalue(py::module& m) {
    m.def("tc_out_mat_rvalue", &carma::tests::test_tc_out_mat_rvalue, "Test type caster");
}

void bind_test_tc_out_row(py::module& m) {
    m.def("tc_out_row", &carma::tests::test_tc_out_row, "Test type caster");
}

void bind_test_tc_out_row_rvalue(py::module& m) {
    m.def("tc_out_row_rvalue", &carma::tests::test_tc_out_row_rvalue, "Test type caster");
}

void bind_test_tc_out_col(py::module& m) {
    m.def("tc_out_col", &carma::tests::test_tc_out_col, "Test type caster");
}

void bind_test_tc_out_col_rvalue(py::module& m) {
    m.def("tc_out_col_rvalue", &carma::tests::test_tc_out_col_rvalue, "Test type caster");
}

void bind_test_tc_out_cube(py::module& m) {
    m.def("tc_out_cube", &carma::tests::test_tc_out_cube, "Test type caster");
}

void bind_test_tc_out_cube_rvalue(py::module& m) {
    m.def("tc_out_cube_rvalue", &carma::tests::test_tc_out_cube_rvalue, "Test type caster");
}

File: headers/carma/tests/test_type_caster.py
"""Test armadillo matrix to numpy array type caster."""
import numpy as np

import test_carma as carma


def test_type_caster_in_mat():
    """Test type caster input handling of matrix."""
    sample = np.random.normal(size=(25, 2))
    npsum = sample.sum()
    accu = carma.tc_in_mat(sample)
    assert np.isclose(accu, npsum)

    sample = np.random.normal(size=(2, 25))
    npsum = sample.sum()
    accu = carma.tc_in_mat(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_in_row():
    """Test type caster input handling of row."""
    sample = np.random.normal(size=(25))
    npsum = sample.sum()
    accu = carma.tc_in_row(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_in_row_2d():
    """Test type caster input handling of 2d row."""
    sample = np.random.normal(size=(1, 25))
    npsum = sample.sum()
    accu = carma.tc_in_row(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_in_col():
    """Test type caster input handling of column."""
    sample = np.random.normal(size=(25))
    npsum = sample.sum()
    accu = carma.tc_in_col(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_in_col_2d():
    """Test type caster input handling of 2d column."""
    sample = np.random.normal(size=(25, 1))
    npsum = sample.sum()
    accu = carma.tc_in_col(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_in_cube():
    """Test type caster input handling of cube."""
    sample = np.random.normal(size=(25, 2, 3))
    npsum = sample.sum()
    accu = carma.tc_in_cube(sample)
    assert np.isclose(accu, npsum)


def test_type_caster_out_mat():
    """Test type caster output handling of matrix."""
    sample = np.random.normal(size=(25, 2))
    mat = carma.tc_out_mat(sample)
    assert np.allclose(mat, sample + 1)


def test_type_caster_out_mat_rvalue():
    """Test type caster output handling of matrix rvalue."""
    sample = np.random.normal(size=(25, 2))
    mat = carma.tc_out_mat_rvalue(sample)
    assert np.allclose(mat, sample + 1)


def test_type_caster_out_row():
    """Test type caster input handling of matrix."""
    sample = np.random.normal(size=(25))
    mat = carma.tc_out_row(sample)
    assert np.allclose(mat, 1 + sample)


def test_type_caster_out_row_2d():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(1, 25)), order='F')
    mat = carma.tc_out_row(sample)
    assert np.allclose(mat, 1 + sample)


def test_type_caster_out_row_rvalue():
    """Test type caster input handling of matrix."""
    sample = np.random.normal(size=(25))
    mat = carma.tc_out_row_rvalue(sample)
    assert np.allclose(mat, 1 + sample)


def test_type_caster_out_col():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(25)), order='F')
    mat = carma.tc_out_col(sample)
    assert np.allclose(mat.flatten(), 1 + sample)


def test_type_caster_out_col_2d():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(25, 1)), order='F')
    mat = carma.tc_out_col(sample)
    assert np.allclose(mat, 1 + sample)


def test_type_caster_out_col_rvalue():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(25)), order='F')
    mat = carma.tc_out_col_rvalue(sample)
    assert np.allclose(mat.flatten(), 1 + sample)


def test_type_caster_out_cube():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(25, 3, 2)), order='F')
    mat = carma.tc_out_cube(sample)
    sw_mat = np.moveaxis(mat, [0, 1, 2], [2, 0, 1])
    assert np.allclose(sw_mat, 1 + sample)


def test_type_caster_out_cube_rvalue():
    """Test type caster input handling of matrix."""
    sample = np.asarray(np.random.normal(size=(25, 3, 2)), order='F')
    mat = carma.tc_out_cube_rvalue(sample)
    sw_mat = np.moveaxis(mat, [0, 1, 2], [2, 0, 1])
    assert np.allclose(sw_mat, 1 + sample)
